### DISCUSSION TITLE
Send Mirror Data To juggluco

### DISCUSSION THREAD

**Discussion by kfishermail** [j-kaltes/Juggluco#66] [Q&A]
I am developing a device to log when an insulin injection occurs.  I know I can manually
add the time/date with juggluco GUI.
I want to automatically add it to JG.  I think the best approach is to send a TCP message to 
port 8795 (mirror port).

Is the format of the mirror messages defined anywhere?
I have downloaded JG source, but I can not find where these messages are being generated.
What source files generate these mirror messages? and where are they parsed?

**Comment by j-kaltes** [comment]
﻿You can best look at the command line version of Mirror (https://www.juggluco.nl/Juggluco/cmdline), because it contains only the relevant files. This is an old version, but you can create a new source by going to an empty directory and run 
`sh $Jugglucodir/Common/src/main/cpp/makecmd.sh 5.1.5`
You shouldn’t do the sending yourself, but save that data locally and rely on the sending mechanism of the mirror system. So the only thing you have to do is take over the command line mirror program and save your data in this program.
The amounts are saved with numsave (nums/numdata.h).
If you run an existing Juggluco (command line or Android) two Numdata structures are present in vector<Numdata*> numdatas. The first one is for data entered in the Juggluco app, the second for data entered in Kerfstok on Garmin watches. See Common/src/main/cpp/cmdline/main.cpp
I could give more specific advice if  I understood what you are trying to do.

**Reply by kfishermail** [reply]
I downloaded JL source from github, and Tried to create new source, but getting errors:
pi@hplaptop:~/myjugg$ sh /home/pi/Juggluco-primary/Common/src/main/cpp/makecmd.sh 5.1.5
../Juggluco-primary/Common/src/main/cpp/makecmd.sh: 8: cannot create /cmdline/version.h: Directory nonexistent
Usage

  cmake [options] <path-to-source>
  cmake [options] <path-to-existing-build>
  cmake [options] -S <path-to-source> -B <path-to-build>

Specify a source directory to (re-)generate a build system for it in the
current working directory.  Specify an existing build directory to
re-generate its build system.

Run 'cmake --help' for more information.

make: *** No rule to make target 'cmdlinesrc.tar'.  Stop.
tar: cmdlinesrc.tar: Cannot open: No such file or directory
tar: Error is not recoverable: exiting now
grep: /CMakeLists.txt../Juggluco-primary/Common/src/main/cpp/makecmd.sh: 14: cannot create home/jka/src/android/Glucodata/Common/src/main/cpp/CMakeLists.txt: Directory nonexistent
: No such file or directory
cp: cannot stat '/README': No such file or directory
../Juggluco-primary/Common/src/main/cpp/makecmd.sh: 16: cd: can't cd to home/jka/src/android/Glucodata/Common/src/main/cpp/..
mv: cannot stat 'cpp': No such file or directory
tar: juggluco-server: Cannot stat: No such file or directory
tar: Exiting with failure status due to previous errors
pi@hplaptop:~/myjugg$


pi@hplaptop:~/myjugg$ ls -l /home/pi/Juggluco-primary/Common/src/main/cpp/makecmd.sh
-rw-r--r-- 1 pi pi 584 Jul 18 03:41 /home/pi/Juggluco-primary/Common/src/main/cpp/makecmd.sh
pi@hplaptop:~/myjugg$

**Reply by j-kaltes** [reply]
Then take the old version. This script only copies the relevant files out of the source of Juggluco. Not much is changed. The script doesn't work because the $gc environmental variable is only defined on my computer. You can also add something like export gc=/home/pi/Juggluco-primary/Common/src/main/cpp
